### ISSUE TITLE
Run-time performance improvement suggestions

### ISSUE THREAD

**Issue by luk036** [luk036/cyclic_quorum#4]
# Run-time performance improvement suggestions

Analysis of the core solvers (`src/diff_cover.cpp`, `src/RL_mt.cpp`, `src/RL_mt2.cpp`, `src/diff_cover3.cpp`, `src/ThreadPool.h`). Suggestions are prioritized by expected impact.

---

## 0. Biggest issue: `RL_mt.cpp` never learns (100% wasted compute)

In `src/RL_mt.cpp` the gradient arrays are allocated and passed to `update()`, but **never written**:

```cpp
// lines 332-337: gradients allocated, zero-initialized
std::vector<std::vector<float>> gradW1(HIDDEN_SIZE1, std::vector<float>(2*N, 0.0f));
...
// lines 340-349: computes gradLogits into a LOCAL vector, then discards it
for (size_t t = 0; t < states.size(); ++t) {
    ...
    std::vector<float> gradLogits(N, 0.0f);   // <-- local, never backpropagated
    ...
}
// line 352: update subtracts 0 * lr from every weight
policyNet.update(gradW1, gradB1, gradW2, gradB2, gradW3, gradB3);
```

There is no backpropagation at all — `update()` subtracts zero. The network stays at its random initialization forever, so `RL_mt.cpp` is effectively **random search with a fixed random policy**. Every episode is wasted work.

`RL_mt2.cpp` already fixes this with real backpropagation, local gradient accumulation, entropy regularization, and batched updates.

**Action:** Treat `RL_mt2.cpp` as the RL baseline; either delete `RL_mt.cpp` or port the missing backprop into it.

---

### 1a. Add bracelet symmetry breaking (algorithmic, biggest win)

`diff_cover3.cpp` adds a `q[]` array + `CheckRev()` (lines 96-102, 190-201) to prune reflection-equivalent branches. The main `diff_cover.cpp` uses only the `b[]`/`min` check in `PrintD` — no reversal symmetry breaking. Since reflection maps every solution to another solution, this can cut the search tree substantially (roughly up to ~2x on the canonical-class frontier, more at deeper levels). Port the `q`/`CheckRev` machinery from `diff_cover3.cpp`.

### 1b. Finer-grained work decomposition / load balance

`InitParallel` (lines 216-242) splits work **only on `a[1]`**, giving ~`N/2 - N/D` tasks (e.g. 47 for N=112, D=12). Those subtrees are wildly different sizes, so with 12 workers there is tail imbalance: one worker can own a huge subtree while others idle. The `ThreadPool` queue already provides dynamic assignment, so the fix is **granularity**, not scheduling:

- Split at `a[2]` (or depth 3) and enqueue those subproblems, so workers pull many small units. Restructure `Gen11` to emit its first-level children as tasks.
- Alternatively use a work-stealing deque so a busy worker's remaining range can be stolen.

### 1c. Output I/O

`PrintD` calls `printf` + `fflush(stdout)` **per solution** (lines 144-149). For large N with many solutions this serializes on stdout inside the hot path. Accumulate results per worker into a local buffer/string and write once, or at minimum drop the per-result `fflush`.

### 1d. Build flags / PGO

Already `-O3`. Add for this branch-heavy backtracking:

```
-march=native -funroll-loops -flto
```

and, best value, **PGO** (`-fprofile-generate` -> run a representative `N,D` -> `-fprofile-use`). Branch-heavy recursive search typically gains 10-20%. Also consider `-fno-exceptions` on the search TU if the thread pool permits.

### 1e. Micro (lower value)

- `step_forward`/`step_backward` (lines 95-124): the `min` is already branchless. `a[]` is <=17 ints (L1-resident), `differences[]` <=121 bytes — memory behavior is already good. SIMD is impractical because of the scatter into `differences[diff]`.
- `__builtin_expect` on the pruning branch `if (count >= N1 + t*t_1/2)` (line 165) to keep the hot path straight-line.
- The `results`/`future`/`packaged_task` overhead is one-time per top-level task (~47), negligible.

---

### 2a. Don't recompute forward passes in the backward loop

The rollout already calls `forward` per step (line 262), then lines 324-325 call `forward` **again** for every stored state. Cache the `ForwardResult` inside `StepData` during rollout and reuse it. This removes `T = D-1` full forward passes per episode (~half the network cost).

### 2b. Eliminate per-call heap allocations

`forward()` allocates 5 `std::vector`s per call (lines 92-121); `backward()` allocates 3 (lines 137, 150, 166); `softmax()` allocates 1. At millions of episodes x 2T calls this is a lot of allocator traffic. Use thread-local scratch buffers (or fixed-size `std::array` since `HIDDEN_SIZE1/2` are compile-time and `N <= D(D-1)+1` bounds the rest). Also hoist `chosen`/`residues`/`chosenList` out of the episode loop and reuse, and `trajectory.reserve(D-1)`.

### 2c. Vectorize the matvec

The forward pass is a dense `W*x` (256x2N, 128x256, Nx128). It auto-vectorizes with `-O3 -march=native`; add `-ffast-math` for the RL TU (safe here) and consider `#pragma omp simd` or a tiny AVX/FMA GEMM. For N=112, `W1` is ~229 KB — L2-resident, so a cache-blocked layout (`W1` is already row-major, good) will help.

### 2d. Replace the global update mutex with lock-free (Hogwild)

`update()` (lines 183-193) holds `networkMutex` while writing ~104K floats, and `forward()` reads the same weights **without** the lock — both a contention point and a data race (torn reads). The standard fix for this workload is **Hogwild! / async SGD**: drop the mutex, accept benign races on weight updates. It removes all lock contention and scales near-linearly with threads. Increase `LOCAL_UPDATE_INTERVAL` to batch more before touching shared weights.

### 2e. Sampling

`std::discrete_distribution` reconstructs an internal table every step. Since normalized `probs` are already available, sample with a manual cumulative sum + one `uniform_real` (O(N), no allocation).

### 2f. Search-quality (not raw speed, but time-to-solution)

- **Cache/prune by symmetry**: fix element 0 (done) and additionally fix the reflection canonical form, as in the combinatorial solver.
- **Better state**: the `2N` binary vector (chosen + residues) is sparse; a representation keyed on uncovered residues only would shrink the input layer and speed `W1*x`.
- **Reward shaping**: potential-based shaping on uncovered-residue count usually converges faster than raw `newCovered`.
- **Population/periodic sync** instead of a single shared net avoids all threads collapsing onto an identical policy.

---

## 3. Measure first

There is no profiling in the repo. Before optimizing, get a baseline:

- **Windows**: VS `PerfView` / `WPA`, or `perf` under WSL.
- Quick and dirty: compile with `-pg`/`gprof`, or sample with `-fno-omit-frame-pointer` + a profiler.
- Instrument episode throughput (episodes/sec) for RL and nodes/sec for the backtracker so each change is measurable.

---

## Suggested order of attack

1. `RL_mt.cpp` correctness (stop wasting 100% of RL compute) — or standardize on `RL_mt2.cpp`.
2. `diff_cover.cpp`: port bracelet symmetry breaking from `diff_cover3.cpp`.
3. `RL_mt2.cpp`: cache forward results + remove allocations + Hogwild.
4. Finer work decomposition in `InitParallel`.
5. PGO / build flags.

**Comment by luk036** [comment]
## Implementation progress

Implemented so far (branch working tree, not yet committed):

### 0. `RL_mt.cpp` — was not learning at all (FIXED)
The gradient arrays were allocated and passed to `update()`, but never written — `gradLogits` was computed into a local and discarded, so `update()` subtracted zero. The network stayed at its random init forever (effectively random search).

Ported real REINFORCE backprop:
- Added `ForwardResult` (pre/post-ReLU activations) and `PolicyNetwork::backward()`.
- `workerThread` now stores per-step sampling distributions and accumulates gradients before `update()`.
- Added a correctly-signed entropy bonus (`ENTROPY_BETA = 0.01`) to prevent policy collapse.

Verified: throwaway instrumented build shows average coverage rising monotonically with `ENTROPY_BETA=0` (13.77 -> 13.91 over 24k episodes); solutions now found (e.g. episode 20 vs 1107 for the old fixed-random policy on one N=31,D=6 run).

### 1. `diff_cover.cpp` — bracelet (reflection) symmetry breaking (DONE)
Ported the `q[]` membership array + `CheckRev()` from `diff_cover3.cpp` (`GenD`->`BraceFD`, `Gen11`->`BraceFD11`).

Verified:
- `diff_cover3` confirmed a correct reference first (identical, all-valid solution sets vs original across N=15/21/31/57/73/91).
- New `diff_cover.cpp` is output-identical to that reference on all those cases.
- Speedup: ~1.08x (N=73,D=9), ~1.15x (N=91,D=10). Modest.

### 2. `RL_mt2.cpp` — entropy sign was inverted (FIXED)
`gradLogits` subtracted the entropy term, i.e. it *penalised* entropy and encouraged collapse. Changed to `+`. On N=31,D=6 this went from timing out on every trial to finding valid solutions at episodes 141 / 914.

### 2a/2b/2d. `RL_mt.cpp` — RL efficiency items (DONE)
- **2a**: cache `ForwardResult` during rollout; backward reuses it (no `T` extra forward passes per episode).
- **2b**: hoisted all per-thread buffers out of the episode loop; `softmax` writes into a buffer.
- **2d**: removed the update mutex (Hogwild-style lock-free async SGD).

Measured (40 000 episodes, N=31,D=6):
- 1 thread: 75.3s -> 50.8s (**1.48x**)
- 10 threads: 5.34s (**9.5x** scaling)
- Outputs validated as genuine difference covers.

### Still open (deferred)
- 2c: SIMD/BLAS for the dense matvec (nested-`std::vector` indexing currently blocks auto-vectorisation).
- 2e: replace `std::discrete_distribution` with manual inverse-CDF sampling.
- 2f: symmetry pruning / state representation / reward shaping for faster convergence.
- 1b: finer-grained work decomposition in `InitParallel`.
- 1d: PGO / `-march=native -funroll-loops -flto`.
